Add Debian/Ubuntu packaging scripts

chenxiaolong opened the pull request:
See: #22

chenxiaolong committed:
Add Debian/Ubuntu packaging scripts

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chrishoage commented on dist/build_source_package.sh:
```diff
@@ -109,6 +109,37 @@ build_pkgbuild() {
 }
 
+# Build deb source package for Debian/Ubuntu
+build_debsrc() {
```
this is super pedantic, sorry, but we're creating a dsc which can be used to create a deb.

chenxiaolong replied:
Good call. Fixed!

chrishoage reviewed (approved):
`dh-exec` is also required. I'm not sure if I should have had it included with the other deb tools but I ran into this error while running the script 

```
dpkg-checkbuilddeps: error: Unmet build dependencies: dh-exec
```

However it worked! 

```
› sudo dpkg -i ipmi-fan-control_0.3.0.r11.gitf9ea589-1_amd64.deb
(Reading database ... 126986 files and directories currently installed.)
Preparing to unpack ipmi-fan-control_0.3.0.r11.gitf9ea589-1_amd64.deb ...
Unpacking ipmi-fan-control (0.3.0.r11.gitf9ea589-1) over (0.3.0) ...
Setting up ipmi-fan-control (0.3.0.r11.gitf9ea589-1) ...
```

it even handled the config files correctly during an upgrade.  

```
› cat /etc/ipmi-fan-control.toml.dpkg-dist 
# Definition of a logical fan zone.
[[zones]]
# IPMI session. If unspecified, the `default` session is used, which runs
# ipmitool with no arguments. Sessions are defined below in the `sessions`
```

chenxiaolong commented:
> `dh-exec` is also required. I'm not sure if I should have had it included with the other deb tools but I ran into this error while running the script

`dh-exec` is specified under `Build-Depends` in `dist/debian/control`, but I don't think debuild will automatically install build dependencies. Unless I'm missing something, I think those have to be manually installed before building the package.

EDIT: Looks like `mk-build-deps` can parse `debian/control` and install the required build dependencies: https://serverfault.com/questions/127625/given-a-debian-source-package-how-do-i-install-the-build-deps

chrishoage commented:
Good to know about `mk-build-deps` thanks!

Edit: Thinking on this a little more, `dh-exec` was required simply to run `./build_source_package.sh -t dsc` at all, so it's a bit of a chicken and egg problem to know if deps are missing

chenxiaolong commented:
> Edit: Thinking on this a little more, `dh-exec` was required simply to run `./build_source_package.sh -t dsc` at all, so it's a bit of a chicken and egg problem to know if deps are missing

Huh, that's interesting. I was able to reproduce that in a fresh `ubuntu:20.04` docker container. It looks like the reason `debuild` requires the build deps even for making the source package is because it runs `./debian/rules clean`, which could theoretically do anything that requires the build deps.

Would it be worth updating `check_tools dch debuild` in `build_source_package.sh` to:
```
check_tools cargo dch debuild dh-exec
```
so that it fails faster?

chrishoage commented:
> so that it fails faster?

That's what I was thinking if it wasn't something silly that I did to myself. Looks like it wasn't me, it was Ubuntu :slightly_smiling_face:

chenxiaolong commented:
OK cool, I'll update the `check_tools` line then.

---

Re. uploading to launchpad.net, it's more complicated than I thought:

* The target distro must be listed in `debian/changelog`
* The version number must also be different for every distro version because the packages are all placed in the same directory in the repo (but with distro-specific metadata that picks the correct `.deb`s to use)

I have to build multiple source packages with these in `debian/changelog`:
```
# Local builds
 ipmi-fan-control (0.3.0.r8.git2d545e3-1) UNRELEASED; urgency=medium
# Ubuntu 18.04 PPA
 ipmi-fan-control (0.3.0.r8.git2d545e3-1~ubuntu18.04) bionic; urgency=medium
# Ubuntu 20.04 PPA
 ipmi-fan-control (0.3.0.r8.git2d545e3-1~ubuntu20.04) focal; urgency=medium
# Ubuntu 20.10 PPA
 ipmi-fan-control (0.3.0.r8.git2d545e3-1~ubuntu20.10) groovy; urgency=medium
```

I think I can update the `build_source_package.sh` script to work like:
```
# Local builds
./dist/build_source_package.sh -t dsc
# Uploadable builds
./dist/build_source_package.sh -t dsc -d bionic
./dist/build_source_package.sh -t dsc -d focal
./dist/build_source_package.sh -t dsc -d groovy
```

One thing I'll have to figure out is how to determine the version suffix (`~ubuntu20.10`) from the version name (eg. `groovy`). The launchpad docs explicitly discourage using a version suffix like `~groovy`. I don't want to hardcode a map of version numbers into the script.

---

EDIT: On second thought, maybe I just do:
```
# Local builds
./dist/build_source_package.sh -t dsc
# Uploadable builds
./dist/build_source_package.sh -t dsc -d bionic -s '~ubuntu18.04'
./dist/build_source_package.sh -t dsc -d focal -s '~ubuntu20.04'
./dist/build_source_package.sh -t dsc -d groovy -s '~ubuntu20.10'
```
That would go in the GitHub Actions build pipeline. The script wouldn't need to know about distros versions as all.

chenxiaolong committed:
build_source_package.sh: Add check for cargo and dh-exec to fail faster if build deps are not installed

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chrishoage commented:
>  it's more complicated than I thought

Sounds like Debian distros in a nutshell :wink: 

> I don't want to hardcode a map of version numbers into the script.

Depending on how you were planning on building these you could either pass in the version (e.g. `-d focal -dv 20.04`?

>  Edit: On second thought

Edit: Looks like we had the same thought!

chenxiaolong commented:
> Depending on how you were planning on building these you could either pass in the version (e.g. -d focal -dv 20.04?

Ha! I just had the same idea 😄 I'll do that then.

chenxiaolong committed:
dist/build_source_package.sh: Add support for specifying the dsc target distro and version suffix

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
dist/build_source_package.sh: Ensure *_source.{build,buildinfo,changes} files are saved

This are required for launchpad.net uploads.

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong commented:
Hit another roadblock :( Looks like launchpad.net will not accept unsigned source packages.

I'm not comfortable with uploading my own GPG key anywhere, so I think I need to:

1. Create a launchpad.net bot account.
2. Create an `ipmi-fan-control` team in launchpad because individual PPAs don't allow granting permissions to other users.
3. Create a new GPG key for the bot account and register the public key in launchpad.net and the private key in this repo's GitHub Actions secrets.
4. Update `build_source_package.sh` to not set `DEBFULLNAME` and `DEBEMAIL` so that the bot account's name and email can be used.
5. Update `build_source_package.sh` to make it possible to run `debuild` without `-us -uc`.

---

I wonder if it's even worth it to use launchpad.net. It feels like a pain to do anything automated with it.

There's also https://build.opensuse.org/, where a single SRPM + dsc + PKGBUILD can be uploaded and it'll build packages and create repos for all supported versions of Fedora/CentOS/Debian/Ubuntu/Arch.

chrishoage commented:
> I'm not comfortable with uploading my own GPG key anywhere

Will they allow a [subkey ](https://wiki.debian.org/Subkeys) to be uploaded instead? 

> I wonder if it's even worth it to use launchpad.net. It feels like a pain to do anything automated with it.

Hey now you're starting to sound like me :wink: 

Honestly, with something that gets updated so infrequently it seems fine to just have people install directly :woman_shrugging: 

> There's also https://build.opensuse.org/, where a single SRPM + dsc + PKGBUILD can be uploaded and it'll build packages and create repos for all supported versions of Fedora/CentOS/Debian/Ubuntu/Arch.

Perhaps I'm not looking in the right place, but it doesn't seem to create a repo for Debian based distros

>The Build Service will not create a source repository for .deb-based distributions; only the .deb file will be created and published.

https://en.opensuse.org/openSUSE:Build_Service_Debian_builds#Package_repositories

chenxiaolong commented:
> Hey now you're starting to sound like me 😉

:joy:

> Will they allow a [subkey ](https://wiki.debian.org/Subkeys) to be uploaded instead?

I didn't think they supported that, but the last time I checked was probably 8 years ago. It looks like they added support just a few months after I tried to use it: https://bugs.launchpad.net/launchpad/+bug/1053568

> Perhaps I'm not looking in the right place, but it doesn't seem to create a repo for Debian based distros
> 
> > The Build Service will not create a source repository for .deb-based distributions; only the .deb file will be created and published.
> 
> https://en.opensuse.org/openSUSE:Build_Service_Debian_builds#Package_repositories

I think "source repository" might be referring to `apt-get source`. You'd be able to `apt-get install ipmi-fan-control` normally, but you couldn't use `apt-get source ipmi-fan-control` to download the dsc and build locally (though of course, you could still curl/wget those files from build.opensuse.org).

I'll test it out over the next couple days and see how well it works. If the service is good, then I'll probably use it. We'd get Debian packages for "free" then, as opposed to just Ubuntu on launchpad.

---

EDIT: I don't think launchpad is an option. It doesn't appear to allow network access during the build.

chenxiaolong committed:
dist/build_source_package.sh: Add all dependencies into generated tarball

This allows services like launchpad.net and build.opensuse.org that do
not allow internet access to build this project.

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
debian/patches: Add patches to support Rust 1.41

The minimum supported Rust version for both ipmi-fan-control and the
tokio dependency are 1.45, but Debian 10 and Ubuntu 20.04 only ship with
Rust 1.41. Thankfully, it's very easy to patch things to work with the
older Rust version.

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
dist/build_source_package.sh: Add support for creating signed dsc/changes files

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
rpm/ipmi-fan-control.spec.in: Add workaround for OBS Fedora builds

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong commented:
So build.opensuse.org seems to work incredibly well. I set up a test repo to try things out:

* Repo page: https://build.opensuse.org/package/show/home:chenxiaolong:test/ipmi-fan-control
* User friendly landing/setup page: https://software.opensuse.org//download.html?project=home%3Achenxiaolong%3Atest&package=ipmi-fan-control

I was able to successfully build packages for CentOS/Fedora/OpenSUSE/Debian/Ubuntu with a single RPM and (unsigned) dsc.

I had to make some changes to this project to get things to work though:
* build.opensuse.org also does not allow internet access during the build, so I updated `./dist/build_source_package.sh -t tarball` to run `cargo vendor` so that all Rust dependencies' source code is bundled into the tarball.
* The Debian 10 and Ubuntu 20.04 repos only have Rust 1.41, which is too old, so I added some backport patches to make the compilation work.

---

The only thing to figure out is whether it's worth it to add a binary repo for Arch. OBS accepts the PKGBUILD as you have it now (because it references the tarball directly and thus, does not require internet). Maybe the AUR is good enough for Arch--3rd party binary repos seem to be uncommon there.

chrishoage commented:
> I set up a test repo to try things out

I will setup the apt source this weekend and try it out. 

> Maybe the AUR is good enough for Arch--3rd party binary repos seem to be uncommon there.

While this is true, it's not so unheard of that it doesn't happen. In fact the Arch Wiki has 4 opensuse repos listed in the [unofficial user repositories](https://wiki.archlinux.org/index.php/unofficial_user_repositories) page.

If I were in your position I would use opensuse for everything since it is far simpler to manage everything in one place. 

The nice thing about the AUR is that it's a _user_ repository. Let someone else maintain the PKGBUILD there if they desire :slightly_smiling_face:

chenxiaolong commented:
Sounds good to me! In that case, I'll keep the PKGBUILD as is and update `build_source_package.sh` to use `sed` if `updpkgsums` is unavailable.

chenxiaolong committed:
dist/build_source_package.sh: Replace updpkgsums with sha256sum + sed

updpkgsums is unfortunately only available on Arch and GitHub Actions
runs on Ubuntu.

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
dist/build_source_package.sh: Follow distro-specific versioning scheme for git packages

For stable versions (eg. 0.3.0):

| Type     | Version | Release       |
|----------|---------|---------------|
| RPM      | 0.3.0   | 1.fc33        |
| DEB      | 0.3.0   | 1 or 1ubuntu1 |
| PKGBUILD | 0.3.0   | 1             |

For git versions (eg. 0.3.0.r19.git0b6d53a):

| Type     | Version              | Release               |
|----------|----------------------|-----------------------|
| RPM      | 0.3.0                | 1.r19.git0b6d53a.fc33 |
| DEB      | 0.3.0+r19+git0b6d53a | 1 or 1ubuntu1         |
| PKGBUILD | 0.3.0.r19.git0b6d53a | 1                     |

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
dist/build_source_package.sh: Make the tarball creation reproducible

This follows the recommendation from [1] and should result in the
tarball being byte-for-byte reproducible no matter which system it is
created on.

This forces the following changes to the tarball:

* Sorts all entries inside the tarball (no longer dependent on
  filesystem traversal order)
* UID/GID is set to 0 and no owner/group username is stored
* The modification time is set to 0 (unix epoch time)
* The exthdr.name PAX header is set to exclude the PID of the tar
  process
* The file access and creation times are deleted

[1] https://reproducible-builds.org/docs/archives/

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
.github/workflows: Build all source package types

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
dist/build_source_package.sh: Make tarball creation less verbose

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
.github/workflows: Temporarily use ubuntu-20.04 until ubuntu-latest is updated

debhelper >= 12 is needed, but the GitHub Actions' /etc/apt/sources.list
is set up in a way that debhelper doesn't get installed from the
bionic-backports repo.

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
PKGBUILD.in: Temporarily work around zstd-compressed packages not being signed on OBS

See: https://github.com/openSUSE/open-build-service/pull/10570

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
.github/workflows: Set up automated OBS upload for release builds

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
.github/workflows/release.yml: Use 0.0.0.<branch> when testing the release workflow

The old scheme (0.0.0-<branch>) trips up dist/build_source_package.sh,
which ascribes meaning to dashes.

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
ipmi-fan-control.spec: Use Recommends for smartmontools where possible

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chenxiaolong committed:
dist: Use consistent package summaries and descriptions on all distros

Signed-off-by: Andrew Gunnerson <chillermillerlong@hotmail.com>

chrishoage commented:
Tried out the ubuntu test repo today and it appeared to work great!

chenxiaolong commented:
Awesome! I tested the remaining distros via docker and a remote IPMI session and those seem to work well too. Arch is the only exception, where having `TERM=xterm` set causes ipmitool to print out `\e[?2004l` and `\e[?2004h` sequences (apparently part of "bracketed paste mode" which I've never heard of), which trips up ipmi-fan-control. It's not a problem when starting up via systemd though.

I think this PR should be ready to go. I'll merge this, make a 0.3.1 release (no functional changes) so packages get built, and then update the README to add information about the new packages.
